### PR TITLE
Feature/websocket

### PR DESCRIPTION
**Pull request by LuisGP**
Depends on #11 

This PR migrates WebSocket to make use of xtypes.

### CODE REVIEW

File: packages/websocket/src/Client.cpp [hunk 1]
```diff
@@ -289,15 +289,15 @@ class Client : public Endpoint
...
+      const xtypes::DynamicType& message_type,
       const std::string& id,
       const YAML::Node& configuration) override
   {
     if(_connection)
     {
       _connection->send(
             get_encoding().encode_advertise_msg(
-              topic, message_type, id, configuration));
+                topic, message_type.name(), id, configuration));
```
Comment (lemunozm) [inline packages/websocket/src/Client.cpp]:
```suggestion
              topic, message_type.name(), id, configuration));
```

File: packages/websocket/src/Encoding.hpp [hunk 2]
```diff
@@ -76,6 +76,15 @@ class Encoding
       const std::string& service_type,
       const std::string& id,
       const YAML::Node& configuration) const = 0;
+
+  virtual bool add_type(
+        const xtypes::DynamicType& type,
+        const std::string& type_name = "")
+  {
```
Comment (lemunozm) [inline packages/websocket/src/Encoding.hpp]: Should this have a body or should be removed?

File: packages/websocket/src/Endpoint.cpp [hunk 3]
```diff
@@ -100,16 +117,16 @@ bool Endpoint::configure(
...
     const std::string& topic_name,
-    const std::string& message_type,
+    const xtypes::DynamicType& message_type,
     SubscriptionCallback callback,
     const YAML::Node& configuration)
 {
   _startup_messages.emplace_back(
         _encoding->encode_subscribe_msg(
-          topic_name, message_type, "", configuration));
+            topic_name, message_type.name(), "", configuration));
```
Comment (lemunozm) [inline packages/websocket/src/Endpoint.cpp]:
```suggestion
          topic_name, message_type.name(), "", configuration));
```

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 4]
```diff
@@ -71,7 +185,22 @@ static std::string get_optional_string(
     const std::string& key)
 {
   const auto it = object.find(key);
-  return it == object.end()? "" : soss::json::to_string(it.value());
+  //return it == object.end()? "" : soss::json::to_string(it.value());
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]: Can be removed?

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 5]
```diff
@@ -71,7 +185,22 @@ static std::string get_optional_string(
...
+  //return it == object.end()? "" : it.value().dump();
+
+  if(it == object.end())
+  {
+      return "";
+  }
+  else
+  {
+      std::string temp_str = it.value().dump();
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]: Remember the 2 spaces for tabs :fearful:

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 6]
```diff
@@ -83,25 +212,36 @@ static std::string get_required_string(
   if(it == object.end())
     throw_missing_key(object, key);
 
-  return soss::json::to_string(it.value());
+  //return soss::json::to_string(it.value());
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]: Remove?

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 7]
```diff
@@ -83,25 +212,36 @@ static std::string get_required_string(
...
+  return json::convert(type, it.value());
 }
 
 //==============================================================================
 class RosbridgeV2_0 : public Encoding
 {
 public:
+  RosbridgeV2_0()
+  {
+      types_ = xtypes::idl::parse(idl_messages).get_all_types();
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]: beautiful

File: packages/websocket/unit-test/CMakeLists.txt [hunk 8]
```diff
@@ -10,6 +10,11 @@ target_link_libraries(soss-websocket-unit-test
     soss-websocket
     yaml-cpp
     OpenSSL::SSL
+    #soss::json
```
Comment (lemunozm) [inline packages/websocket/unit-test/CMakeLists.txt]: Should be removed definitely?

File: packages/websocket/src/Encoding.hpp (hunk 2 above)
Comment (LuisGP) [reply packages/websocket/src/Encoding.hpp]: It's a default implementation. `rosbridge_v2_encoding` overrides it. I don't think it should be mandatory to implement by potential new `Encoding` impl.

File: packages/websocket/src/rosbridge_v2_encoding.cpp (hunk 5 above)
Comment (LuisGP) [reply packages/websocket/src/rosbridge_v2_encoding.cpp]: :fearful:

File: packages/websocket/unit-test/CMakeLists.txt (hunk 8 above)
Comment (LuisGP) [reply packages/websocket/unit-test/CMakeLists.txt]: Absolutely

File: packages/websocket/src/Encoding.hpp (hunk 2 above)
Comment (lemunozm) [reply packages/websocket/src/Encoding.hpp]: ok

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 9]
```diff
@@ -83,25 +210,35 @@ static std::string get_required_string(
   if(it == object.end())
     throw_missing_key(object, key);
 
-  return soss::json::to_string(it.value());
+  std::string temp_str = it.value().dump();
+  if (temp_str.find("\"") == 0)
+  {
+      return temp_str.substr(1, temp_str.size() - 2);
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]:
```suggestion
    return temp_str.substr(1, temp_str.size() - 2);
```

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 10]
```diff
@@ -83,25 +210,35 @@ static std::string get_required_string(
...
+  return json::convert(type, it.value());
 }
 
 //==============================================================================
 class RosbridgeV2_0 : public Encoding
 {
 public:
+  RosbridgeV2_0()
+  {
+      types_ = xtypes::idl::parse(idl_messages).get_all_types();
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]:
```suggestion
    types_ = xtypes::idl::parse(idl_messages).get_all_types();
```

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 11]
```diff
@@ -120,13 +257,23 @@ class RosbridgeV2_0 : public Encoding
 
     const std::string& op_str = op_it.value().get<std::string>();
 
+    xtypes::DynamicType::Ptr type_ptr;
+
+    auto type_it = types_.find(op_str);
+    if (type_it != types_.end())
+    {
+        type_ptr = type_it->second;
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]:
```suggestion
      type_ptr = type_it->second;
```

File: packages/websocket/src/rosbridge_v2_encoding.cpp [hunk 12]
```diff
@@ -309,9 +474,62 @@ class RosbridgeV2_0 : public Encoding
     output[JsonTypeNameKey] = service_type;
     output[JsonServiceKey] = service_name;
 
+    types_by_topic_[service_name] = service_type;
+
     return output.dump();
   }
 
+  const xtypes::DynamicType& get_type(
```
Comment (lemunozm) [inline packages/websocket/src/rosbridge_v2_encoding.cpp]: Check tabs of this sections

**Pull request merged**